### ISSUE TITLE
Question: k8s-bigip-ctlr + openshift operator + NodePort mode + as3 template

### ISSUE THREAD

**Issue by fabriziofiorucci** [F5Networks/k8s-bigip-ctlr#1796]
Hi, I have Openshift 4.6 and CIS installed through its operator (https://clouddocs.f5.com/containers/latest/userguide/openshift/#installing-cis-using-operators-on-openshift-cluster).
My use case is to have CIS configure LTM with a FastL4 VS to balance a set of NGINX Plus KICs running in NodePort mode (ie. no CNI/VXLAN/Calico/etc involved, just plain NodePort).

The whole thing works just fine if I set up k8s-bigip-ctlr manually using manifests, but for some reason going through the Openshift operator doesn't seem to honour the AS3 template (the same AS3 template that runs just fine when k8s-bigip-ctlr is spun up using manifests instead of the operator).

My k8s-bigip-ctlr Deployment manifest args are:

            "--bigip-username=$(BIGIP_USERNAME)",
            "--bigip-password=$(BIGIP_PASSWORD)",
            "--bigip-url=https://x.z.y.t:443",
            "--bigip-partition=cis",
            "--pool-member-type=nodeport",
            "--insecure",
            "--agent=as3",
            "--manage-ingress=false",
            "--as3-validation=true",
            "--node-poll-interval=30",
            "--verify-interval=30",
            "--log-level=DEBUG"

whereas the "F5BigIpCtlr" manifest args as set up by the operator are:

    agent: as3
    as3_validation: true
    bigip_partition: cis
    bigip_url: https://x.z.y.t
    insecure: true
    log_as3_response: true
    log_level: debug
    manage_ingress: false
    pool-member-type: nodeport
    manage_configmaps: true
    share-nodes: true
    namespace-label: use_cis=true

For some reason, k8s-bigip-ctlr pods log reports:

2021/04/19 15:22:11 [ERROR] Ingress IP Address is not provided. Unable to process ingress resources. Either configure controller with 'default-ingress-ip' or Ingress with annotation 'virtual-server.f5.com/ip'.
even though it should honour the "manage_configmaps" argument. After that, AS3 communication with LTM shows:
2021/04/19 15:23:21 [DEBUG] [AS3] Posting AS3 Declaration
2021/04/19 15:23:21 [DEBUG] [AS3] posting request to https://x.z.y.t/mgmt/shared/appsvcs/declare/
2021/04/19 15:23:21 [ERROR] [AS3] Big-IP Responded with code: 422
2021/04/19 15:23:21 [ERROR] [AS3] Raw response from Big-IP: map[code:422 declarationFullId: errors:[/cis/Shared/ingress__80: should have required property 'virtualAddresses'] message:declaration is invalid]

I might be missing something here, but my question is: what args shall I set up in the F5BipIpCtlr yaml file so that the operator-installed k8s-bigip-ctlr can correctly fetch the AS3-template and deploy it on my LTM?

Thanks

**Comment by trinaths** [comment]
Tried as3 configMap with OCP operator installed CIS and see no issues. Please check your environment with any ingress resource which is not proper. Closing this issue.

**Comment by fabriziofiorucci** [comment]
Hello Trinath,

thank you: would you mind sharing the working yaml file used by the operator to spin up CIS and the actual ConfigMap containing the as3 template so that I can check as well?

**Comment by trinaths** [comment]
@fabriziofiorucci The same as3 configMap shared via email. I made no changes.

**Comment by fabriziofiorucci** [comment]
Thank you: one last question, could you just confirm the args used to start k8s-bigip-ctrl in the yaml file used by the operator?

**Comment by fabriziofiorucci** [comment]
Hi, I managed to get it to work. This definitely looks like some kind of bug.

The scenario is:

- k8s-bigip-ctlr running on Openshift 4.5/4.6 using its operator
- CIS must provision an LTM via AS3 (in my case it's a FastL4 vs, but this is irrelevant in the context of the bug) to balance several NGINX KICs running in NodePort mode (this can be any kind of Service running in NodePort mode)

For some reason when k8s-bigip-ctlr is started through the operator using:

agent: as3
as3_validation: true
bigip_partition: cis
bigip_url: https://x.z.y.t
insecure: true
log_as3_response: true
log_level: debug
manage_ingress: false
pool-member-type: nodeport
manage_configmaps: true
share-nodes: true

it does not fetch the as3-template ConfigMap (even though manage_configmaps is set to true) and wants to configure LTM as an ingress. In order to make it work, I needed to add:

manage_routes: false
route_vserver_addr: 1.2.3.4

Doing this CIS gets the as3 template ConfigMap and works just fine.
